File: problems/742/solution.go
package main
const MaxUint = ^uint(0) 
const MinUint = 0 
const MaxInt = int(MaxUint >> 1) 
const MinInt = -MaxInt - 1

type Res struct {
    leaf *TreeNode
    distToLeaf, distToTarget int
    foundTarget bool
}

var shorthestPath int
var shorthestLeaf *TreeNode

func findClosestLeaf(root *TreeNode, k int) int {
    shorthestPath, shorthestLeaf = MaxInt, nil
    
    dfs(root, k)
    
    return shorthestLeaf.Val
}

func dfs(root *TreeNode, k int) Res {
    res := Res{nil, MaxInt, MaxInt, false}
    
    if root == nil {
        return res
    }
    
    left, right := dfs(root.Left, k), dfs(root.Right, k)
    
    if left.leaf == nil && right.leaf == nil {
        res.leaf = root
        res.distToLeaf = 0
    } else {
        if left.distToLeaf <= right.distToLeaf {
            res.distToLeaf = 1 + left.distToLeaf
            res.leaf = left.leaf
        } else {
            res.distToLeaf = 1 + right.distToLeaf
            res.leaf = right.leaf
        }
    }
    
    if root.Val == k {
        res.foundTarget = true
        res.distToTarget = 0
        
        shorthestPath = res.distToLeaf
        shorthestLeaf = res.leaf
    } else if left.foundTarget || right.foundTarget {
        res.foundTarget = true
        
        if left.foundTarget {
            res.distToTarget = 1 + left.distToTarget
        } else {
            res.distToTarget = 1 + right.distToTarget
        }
        
        if res.distToTarget + res.distToLeaf < shorthestPath {
            shorthestPath = res.distToTarget + res.distToLeaf
            shorthestLeaf = res.leaf
        }
    }
    
    return res
}

File: problems/742/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Res {
public:
    TreeNode *leaf;
    int distToLeaf, distToTarget;
    bool foundTarget;
    
    Res(TreeNode *l, int dtl, int dtt, bool ft): leaf{l}, distToLeaf{dtl}, distToTarget{dtt}, foundTarget{ft} {}
};

class Solution {
private:
    int shorthestPath = INT_MAX;
    TreeNode *shorthestLeaf = nullptr;
public:
    int findClosestLeaf(TreeNode* root, int k) {
        Res *res = dfs(root, k);
        
        return shorthestLeaf->val;
    }
    
    Res* dfs(TreeNode *root, int k) {
        Res *res = new Res(nullptr, INT_MAX, INT_MAX, false);
        
        if (not root) {
            return res;
        }
        
        Res *left = dfs(root->left, k), *right = dfs(root->right, k);
        
        if (not left->leaf and not right->leaf) {
            res->leaf = root;
            res->distToLeaf = 0;
        } else {
            res->distToLeaf = 1 + (left->distToLeaf <= right->distToLeaf ? left->distToLeaf : right->distToLeaf);
            res->leaf = left->distToLeaf <= right->distToLeaf ? left->leaf : right->leaf;
        }
        
        if (root->val == k) {
            res->foundTarget = true;
            res->distToTarget = 0;
            
            shorthestPath = res->distToLeaf;
            shorthestLeaf = res->leaf;
        } else if (left->foundTarget or right->foundTarget) {
            res->foundTarget = true;
            res->distToTarget = 1 + min(left->distToTarget, right->distToTarget);
            
            if (res->distToTarget + res->distToLeaf < shorthestPath) {
                shorthestPath = res->distToTarget + res->distToLeaf;
                shorthestLeaf = res->leaf;
            }
        }
        
        return res;
    }
};